AF-61 ProxySSH for AWS and improved functionality

sparshev opened the pull request:
This PR implements full SSH debug access, including:
* Rewriting of the proxyssh implementation to support SCP/Port pass-through
   * Integration tests for proxyssh for SSH/SCP/Port forwarding to ensure functionality will stay with us
* Support for AWS machines SSH connection - now the remote machine can be connected with the private key

## Related Issue

#61 

## Motivation and Context

Have no idea how we can live without that

## How Has This Been Tested?

Manually & Automatically

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
AF-61 Rewrote the proxyssh implementation to support SCP/Port passthrough

Commit:
Completed 2 of 3 integration tests needed for PROXYSSH

Commit:
Added support for keys (generation and usage) and fixed a few flaws:

* Using address for session key potentially could lead to security
  issues, so replaced by SessionID which should have less collisions
* Not storing generated password or key in DB, only hash & key
* Moved to ECDSA instead of RSA since takes less space and farely strong

Commit:
Unified command runAndLog routines to use it in tests

Commit:
Fixed issue with printing raw bytes to the log

Commit:
Added using of real ssh and scp binaries and key2password + key2key

Commit:
Added port forwarding testing to the suit

Commit:
Fixed issues with style

Commit:
Fixed ssh proxy listening in the other tests

svenevs commented on lib/util/cmd.go:
```diff
@@ -0,0 +1,93 @@
+
+	cmd := exec.CommandContext(ctx, path, arg...)
+
+	log.Debugf("%s: Executing: %s %s", sec, cmd.Path, strings.Join(cmd.Args[1:], " "))
```
Is there any concern over secrets getting logged here?  Specifically, things like `docker` credentials?  I looked through the use-cases, I don't _think_ it is a problem.  Just wanted to call out in case it might be.  If it is, it might be easiest to add another parameter `logCall bool` to allow the call-site to disable logging?  Might be a tedious change not worth realizing.

Alternatively, since it's called `RunAndLog`, a different `Run` method that doesn't log could be added.  Then just make a helper method that has this additional `logCalls` parameter, and both `Run` and `RunAndLog` pass-through to this helper function.  It's only necessary if there are locations we are worried about secrets getting dumped to the console, but I couldn't determine if those even exist.

sparshev replied:
If someone runs system as debug - it's their problem, but overall you right, we need to ensure no creds are exposed in the log, but it's a tricky task to solve right away. Created a ticket for that: #88

svenevs replied:
> but it's a tricky task to solve right away

Agreed.  Yeah, this looks good for now, we can deal with it in the future.

svenevs commented on lib/util/cmd.go:
```diff
@@ -0,0 +1,93 @@
+			stderr += fmt.Sprintf("\n--- %s: Command execution attempt %d ---\n", sec, counter)
+			stderr += rerr
+			if counter <= retry {
+				// Give command time to rest
+				time.Sleep(time.Duration(counter) * time.Second)
+				continue
```
I'm a big fan of this code getting consolidated into one module.  I noticed that there were two versions, one that does `time.Duration(counter) * time.Second`, and a hard-coded version

```go
// Give command 5 seconds to rest
time.Sleep(5 * time.Second)
continue
```

Most of the use sites are asking for `3` retries, meaning you wait `1` second, then `2` seconds, then `3` seconds.  It seems like if I asked for `30` retries, this behavior of waiting an increasing amount of time is not desirable in favor of just blanket "wait 5 seconds every time."

Was this the version you intended to keep?

sparshev replied:
I do like the increasing timer - because it allows to do retry faster in the beginning then 5 sec, it's important when your app init needs to retry for example and 150 vs 465 sec is quite a big difference, but we don't use 30 retries in the logic and anyone who want - will need to make sure he understands the logic and how much it will wait. Maybe in the future we will swich to timeout and retries within this timeout - but for now for 5-10 retries it works relatively well.

svenevs replied:
Works for me, I can see value to either version.  Just wanted to make sure you were using the 1/2 implementations you wanted in the final shared version :slightly_smiling_face:

svenevs commented on lib/util/cmd.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+// Runs & logs the executable command
+func RunAndLog(sec string, timeout time.Duration, stdin io.Reader, path string, arg ...string) (string, string, error) {
```
The name of the first parameter `sec` got me pretty confused.  I originally thought it was supposed to mean `seconds`, but it doesn't, it's just a descriptor of the command, e.g., `"NATIVE"`, or `"VMX"`, or `"DOCKER"`.  I think the first parameter should be renamed for added clarity.  Perhaps `description string`?

sparshev replied:
Yeah, it's kind of section, but I suppose I can use some better param name

sparshev replied:
fixed in f7ac492

svenevs commented on lib/util/cmd.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// Will retry on error and store the retry output and errors to return
+func RunAndLogRetry(sec string, retry int, timeout time.Duration, stdin io.Reader, path string, arg ...string) (stdout string, stderr string, err error) {
```
If the name of the `sec` parameter for `RunAndLog`, it should be updated here as well.

sparshev replied:
fixed in f7ac492

svenevs commented on tests/helper/files.go:
```diff
@@ -0,0 +1,124 @@
+)
+
+// CreateRandomFiles will take directory and put there as much random files as you want
+func CreateRandomFiles(dir string, amount int) ([]string, error) {
```
Introducing non-determinism in the tests may result in significant debugging challenges.  It might be advisable to instead have a `CreateFiles` that takes in a map of `filename => file data` pairs.  That would make it so the test needs to explicitly enumerate its expected files, which will likely keep it more accessible long term.

The way its used, `"testfile_"+strconv.Itoa(i)` will result in exact filenames being able to be inferred, but only because the test is "peeking into implementation details" of this method.

sparshev replied:
Random is needed here to make sure any sort of caching is out of question. Random affects just binary data, the file names are quite deterministic. So since we don't care about the binaries within the file (only scp protocol cares about that, which is tested by the other people) it's a good way to keep best of both worlds - other ways are too complicated, so leaving as is until we will face some serious issues.

svenevs replied:
Works for me, yes, the filenames are deterministic and that's what matters.

svenevs commented on tests/helper/ssh.go:
```diff
@@ -0,0 +1,290 @@
+			go func() {
+				io.Copy(f, s) // stdin
+			}()
+			io.Copy(s, f) // stdout
```
Probably obvious ... why doesn't this one (`stdout`) get wrapped in a `go func()`?

sparshev replied:
Because it's not needed to be wrapped - it will execute together with the reverse one and block the pipe until the buffer will be closed. Goroutines are needed to execute in parallel with the main thread and here we need just one additional thread to do both things in parallel.

svenevs commented on tests/helper/ssh.go:
```diff
@@ -0,0 +1,290 @@
+	return io.ReadAll(stdout)
+}
+
+// SCP nowadays uses sftp subsystem with no need for scp binary on the target, so use it directly
```
Not a blocking comment ... just wanted to say this is AWESOME 🚀 🙂

sparshev replied:
Yeah, tests needed some creative approach when it's about ssh & passwords...

svenevs commented on docs/openapi.yaml:
```diff
@@ -1672,23 +1674,41 @@ components:
+          type: string
+          description: >
+            SSH key could be used instead of password to access the system.
+        port:
+          type: integer
+          description: >
+            TCP port to connect with SSH client.
```
Are we sure `integer` is correct here?  From what I could see, all of the tests are using `:0` for their port, which won't cause any issues.  This is something about networking I'm not very knowledgeable about; usually, there's up to four digits allowed.  Specifically for anything less than 1000, should a user be able to supply e.g., `0999` and have it work?  Or is requiring `999` (or `99` rather than `0099`) ok?

If the leading zeros are supposed to be allowed, I think we have to make this `type: string`.  If leading zeros aren't a concern, then we can disregard!

sparshev replied:
Tests are using :0 because this will tell the OS to assign a random not-busy port for binding. And this port in the yaml is to specify where to connect on the target remote system.

And sorry, but I did not get the question at all... IP Port range is uint16 from 1 to 65535 (as per IP specs), but for simplicity I use here regular integer, because openapi don't care about how much bytes is there encoding the number.

As per wikipedia:
> 0 - In programming APIs (not in communication between hosts), requests a system-allocated (dynamic) port

sparshev replied:
But this is made me to think - why ResourceAccess has port in it... And for some reason I added it here and there just for fun) So removed from ResourceAccess at all since not used.

svenevs reviewed (changes requested):
This is **awesome**, way better than the previous implementation by a _landslide_ :partying_face:

I added a number of review comments, some are simple quick-to-address spelling issues.  But a number of the other comments probably should get closed as #wontfix or #irrelevant, just jotted down all the thoughts I had going through the code.  Nice work!

sparshev commented on lib/crypt/crypt.go:
```diff
@@ -47,17 +48,17 @@ const (
-	Salt []byte
-	Hash []byte
+	Algo string     `json:"algo"`
+	Prop properties `gorm:"embedded;embeddedPrefix:prop_" json:"prop"`
```
Then we will need to change `Algo` the same way and that will be nasty and the worse one: backward-incompatible so I would prefer to leave it as is.

svenevs replied:
Oh... yeah, that's not worth the change -- good point!

sparshev commented on lib/crypt/crypt.go:
```diff
@@ -122,3 +123,18 @@ func (h *Hash) IsEqual(input string) bool {
+	}
+	return jsonHash, nil
+}
+func (h *Hash) Deserialize(jsonHash string) error {
```
Hmmm weird why it's not catched by golang-ci... Will need to check, because there is no docs for the interface functions.

sparshev replied:
fixed in f7ac492

sparshev commented on lib/crypt/ssh_key.go:
```diff
@@ -0,0 +1,63 @@
+func GetSSHPubKeyFromPem(privPem []byte) ([]byte, error) {
+	// Decode the PEM block
+	block, _ := pem.Decode(privPem)
+	if block == nil {
+		return nil, fmt.Errorf("Unable to parse PEM block")
+	}
```
Hmmm, this is a tricky one - block type could be almost anything - it's just a message, so I don't want to require additional things from the user who want to use some non-standard pem-encoded key. The decoding will fail on the next decoding of the actual key if something goes wrong, so I think we safe enough here.

sparshev commented on lib/drivers/docker/driver.go:
```diff
@@ -333,22 +334,22 @@ func (d *Driver) Deallocate(res *types.Resource) error {
 	for _, volPath := range cVolumes {
 		if strings.Contains(mounts, volPath) {
-			if _, _, err := runAndLog(5*time.Second, "/usr/bin/hdiutil", "detach", volPath); err != nil {
+			if _, _, err := util.RunAndLog("DOCKER", 5*time.Second, nil, "/usr/bin/hdiutil", "detach", volPath); err != nil {
```
Yeah, for macos implementation it's ok - hdiutil is always here. It's just an optimization where I moved runAndLog to util module, so nothing regarding the drivers logic will be implemented here. In theory you right - we should be able to specify the tools and where they are located. Later when the drivers will be rewritten to support multiple OSes - then will be a great time to make them just like native driver with all those bells and whistles to alter the binaries paths.

sparshev replied:
And the context of the changes is quite important - you don't change small things not related to the goal of the change. In theory you could - but in this case drivers should not be touched other then when necessary. For the drivers better to have a separated PR.

svenevs replied:
Ok sounds good to me.  Yeah I was pretty sure this is a guaranteed exists on macOS, just noticed that in some places there was a config path like there is for `sudo` and `tar`, and some places it's a hard-coded string.  But that makes sense to change in a different PR down the road (it's basically a cosmetic change).

sparshev commented on lib/fish/resource_access.go:
```diff
@@ -48,11 +48,26 @@ func (f *Fish) ResourceAccessDelete(uid types.ResourceAccessUID) error {
+// ResourceAccessSingleUsePasswordHash retrieves the password hash from the database *AND* deletes
+// it. Users must request a new Resource Access to connect again.
+func (f *Fish) ResourceAccessSingleUsePasswordHash(username string, hash string) (ra *types.ResourceAccess, err error) {
 	ra = &types.ResourceAccess{}
-	err = f.db.Where("username = ? AND password = ?", username, password).First(ra).Error
+	err = f.db.Where("username = ? AND password = ?", username, hash).First(ra).Error
+	if err == nil {
+		err = f.ResourceAccessDelete(ra.UID)
+		// NOTE: in rare occasions, `err` here could end up propagating to the
+		// caller with a valid `ra`.  However, see ssh_proxy/proxy.go usage,
+		// in the event that our deletion failed (but nothing else), the single
+		// use connection ultimately gets rejected.
+	}
+	return ra, err
+}
```
Probably you need to re-check the logic... `if err == nil {` - means `if there is no error`.

svenevs replied:
Why yes, you are exactly correct -- I completely red that backwards!  😅

sparshev commented on lib/fish/resource_access.go:
```diff
@@ -48,11 +48,26 @@ func (f *Fish) ResourceAccessDelete(uid types.ResourceAccessUID) error {
+	return ra, err
+}
+
+// ResourceAccessSingleUseKey retrieves the key from the database *AND* deletes it.
+// Users must request a new resource access to connect again.
+func (f *Fish) ResourceAccessSingleUseKey(username string, key string) (ra *types.ResourceAccess, err error) {
```
Yep, exactly the same

svenevs replied:
I read this completely backwards, xref https://github.com/adobe/aquarium-fish/pull/87#discussion_r1878806454

sparshev commented on lib/openapi/api/api_v1.go:
```diff
@@ -259,13 +259,36 @@ func (e *Processor) ResourceAccessPut(c echo.Context, uid types.ResourceUID) err
 		ResourceUID: res.UID,
 		Username:    user.Name,
-		Password:    crypt.RandString(64),
+		// We should not store clear password, so convert it to salted hash
+		Password: string(pwdHash),
+		// Key need to be stored as public key
+		Key: string(pubkey),
```
The issue is that we have different types here and returning bytearray as part of json is not very useful for the end user. So strings they are (because they are strings - just stored in []byte format).

svenevs replied:
Ah ok, makes sense, thanks!

sparshev commented on lib/openapi/api/api_v1.go:
```diff
@@ -259,13 +259,36 @@ func (e *Processor) ResourceAccessPut(c echo.Context, uid types.ResourceUID) err
 	}
 	e.fish.ResourceAccessCreate(&rAccess)
 
+	// Overriding the password and key to return user the actual values one time and forget about
+	rAccess.Password = pwd
+	rAccess.Key = string(key)
```
Yep, we are not storing those in DB - just returning to user so he can keep them. The comment says that we overriding the original values (pubkey/hash that are stored in DB) to return password/privkey to user.

svenevs replied:
Ok, that was not clear to me.  I think updating the comment would make it easier to understand, maybe something like

```go
// Now database has had the hashed credentials stored, restore the original values
// to return to the user so they have access to the correct (un-hashed) credentials.
rAccess.Password = pwd
rAccess.Key = string(key)
```

the current comment (particularly the `one time and forget about` part) did not make it clear _why_ we were resetting the values.  I think the above would help future readers understand better, rather than seeing the comment a few lines above and think there is a mistake.

sparshev replied:
Ok fixed it in daba843 and recalled why I needed port in ResourceAccess - to give user a way to find the port to connect to, so replaced it with Address to prepare for later in-cluster serving the request by another node (for example when it's needed to reach some particular local environment).

sparshev commented on lib/proxyssh/proxy.go:
```diff
@@ -9,304 +9,405 @@
+	log.Debugf("PROXYSSH: %s: Login attempt for user %q.", incomingConn.RemoteAddr(), user)
 
-		go p.serveConnection(conn, p.serverConfig)
+	fishUser, err := p.fish.UserGet(user)
+	if err != nil {
+		log.Errorf("PROXYSSH: Unrecognized user %q", user)
+		return nil, fmt.Errorf("Invalid access")
+	}
```
User - is fish user that requested the resource access. Host (probably you mean target remote instance?) will have a separated set of credentials tied to this fish user by password hash/pubkey stored in the Fish DB. There is no generic anything - the target username is defined in the label authentication field, along with password or key if they are not dynamically generated by the driver - and in AWS case they will be generated.

sparshev replied:
So we have user-pass/key in resourceaccess and then it goes to the Resource (where driver stored the credentials to access this particular machine) and connects to the machine, establishing a bridge between the connected user and remote target.

svenevs replied:
Understood.

sparshev commented on lib/proxyssh/proxy.go:
```diff
@@ -9,304 +9,405 @@
+			}
+		}
+
+		// End the communication between the source and destination.
+		srcChn.Close()
+		dstChn.Close()
```
Yeah, it's working like that, but probably better will be to defer.

sparshev replied:
fixed in f7ac492

sparshev commented on lib/proxyssh/proxy.go:
```diff
@@ -9,304 +9,405 @@
+		log.Errorf("PROXYSSH: %s: The src->dst channel was closed unexpectedly: %v", s.SrcAddr, err)
 	}
+	// Properly closing the channel
+	dstChn.CloseWrite()
+	srcChn.CloseWrite()
```
Those ones was a point where I was unable to figure out why scp is not working properly. So as I found the proper way to close the channels is CloseWrite(), so the system sends the right messages and gracefully wrap-up the connection. And yeah, we don't care much about the errors there - but maybe I can add some debugging here to at least record the errors.

sparshev replied:
fixed in f7ac492

Commit:
Fixed weirdness in the logic and mistakes after review

sparshev commented on lib/proxyssh/proxy.go:
```diff
@@ -9,304 +9,405 @@
-		PasswordCallback: sshProxy.passwordCallback,
+	server := proxySSH{fish: f}
+	server.serverConfig = &ssh.ServerConfig{
+		ServerVersion:     "SSH-2.0-AquriumFishProxy",
```
fixed in f7ac492

sparshev commented on lib/proxyssh/proxy.go:
```diff
@@ -9,304 +9,405 @@
+		dstConfig.Auth = append(dstConfig.Auth, ssh.Password(res.Authentication.Password))
+	}
+
+	// Use use private key if it's set for the Resource
```
fixed in f7ac492

sparshev commented on docs/openapi.yaml:
```diff
@@ -1672,23 +1674,41 @@ components:
       type: object
       description: >
-        Authentication information to enable connecting to the machine.
+        Authentication information to enable connecting to the machine
```
fixed in f7ac492

Commit:
Removed failfast to execute all the tests during CI to have a full picture

svenevs commented on lib/drivers/vmx/util.go:
```diff
@@ -233,15 +230,15 @@ func (d *Driver) disksCreate(vmxPath string, disks map[string]types.ResourcesDis
 			"-size", fmt.Sprintf("%dm", disk.Size*1024),
 		}
-		if _, _, err := runAndLog(10*time.Minute, "/usr/bin/hdiutil", args...); err != nil {
+		if _, _, err := util.RunAndLog("VMX", 10*time.Minute, nil, "/usr/bin/hdiutil", args...); err != nil {
```
This is not going to change in this PR.  xref https://github.com/adobe/aquarium-fish/pull/87#discussion_r1878765033

svenevs commented on lib/drivers/vmx/util.go:
```diff
@@ -233,15 +230,15 @@ func (d *Driver) disksCreate(vmxPath string, disks map[string]types.ResourcesDis
 
 		// Attach & mount disk
-		stdout, _, err := runAndLog(10*time.Second, "/usr/bin/hdiutil", "attach", dmgPath, "-mountpoint", mountPoint)
+		stdout, _, err := util.RunAndLog("VMX", 10*time.Second, nil, "/usr/bin/hdiutil", "attach", dmgPath, "-mountpoint", mountPoint)
```
This is not going to change in this PR.  xref https://github.com/adobe/aquarium-fish/pull/87#discussion_r1878765033

svenevs commented on lib/drivers/vmx/util.go:
```diff
@@ -255,13 +252,13 @@ func (d *Driver) disksCreate(vmxPath string, disks map[string]types.ResourcesDis
 
 		// Detach disk
-		if _, _, err := runAndLog(10*time.Second, "/usr/bin/hdiutil", "detach", devPath); err != nil {
+		if _, _, err := util.RunAndLog("VMX", 10*time.Second, nil, "/usr/bin/hdiutil", "detach", devPath); err != nil {
```
This is not going to change in this PR.  xref https://github.com/adobe/aquarium-fish/pull/87#discussion_r1878765033

svenevs commented on lib/drivers/vmx/util.go:
```diff
@@ -255,13 +252,13 @@ func (d *Driver) disksCreate(vmxPath string, disks map[string]types.ResourcesDis
 
 		// Umount disk (use diskutil to umount for sure)
-		_, _, err = runAndLog(10*time.Second, "/usr/sbin/diskutil", "umount", mountPoint)
+		_, _, err = util.RunAndLog("VMX", 10*time.Second, nil, "/usr/sbin/diskutil", "umount", mountPoint)
```
This is not going to change in this PR.  xref https://github.com/adobe/aquarium-fish/pull/87#discussion_r1878765033

Commit:
Added address to ResourceAccess to simplify user's burden

svenevs reviewed (dismissed):
Pending CI getting fixed, looks good to me! :shipit:

Commit:
Fixed a nasty bug with stored hashes of passwords

Hash is generated with random salt by default and if we don't know
salt (our case) - we can't pull from DB and remove the access quickly.
So instead I store just hash there with empty salt. This works, because
the access is temporary and one-shot, so technically no issues, but
if we will change that at some point - this mechanism is needed to be
reworked.

sparshev commented:
Found a bug thanks to tests - I implemented hashing of the passwords and forgot it stores random salt, so we can't easily generate the same on ssh request without knowing salt of it.

So instead I store just hash there with empty salt. This works, because the access is temporary and one-shot, so technically no issues, but if we will change that at some point - this mechanism is needed to be reworked.

Commit:
Ok, finally figured out the last issue we had with the tests

Apparently the Mac & Lin network layers (or most probably ssh clients of
different versions) are handling the input a bit differently. If you will
easily get away with just buffer as input on Mac and SSH (9.6) client
will not close the src->dst channel - it's not the case with Linux SSH
(9.2p1) client: it will do that and you would never see any response from
the server. So instead we now emulating the ssh terminal input as well
and not closing the input until it's necessary.

Also I found a wrong src->dst messages, so flipped them over and grately
improved the ssh testing where it knocks the mock SSHD first and then
retries the same with proxy. And last thing - found a couple of lost
errors because serve is running as a goroutine and overall increased
verbosity of the mock sshd, so should be much easier to debug now.

sparshev commented:
Apparently the Mac & Lin network layers (or most probably ssh clients of different versions) are handling the input a bit differently. If you will easily get away with just buffer as input on Mac and SSH (9.6) client will not close the src->dst channel - it's not the case with Linux SSH (9.2p1) client: it will do that and you would never see any response from the server. So instead we now emulating the ssh terminal input as well and not closing the input until it's necessary.

Also I found a wrong src->dst messages, so flipped them over and grately improved the ssh testing where it knocks the mock SSHD first and then retries the same with proxy. And last thing - found a couple of lost errors because serve is running as a goroutine and overall increased verbosity of the mock sshd, so should be much easier to debug now.

Commit:
Fixed static checks

Commit:
Added handler for Sftp mock server, otherwise it's failing hard

Commit:
Apparently CI is running with older scp versions, so forcing SFTP

Commit:
Forgot a couple of things
